### PR TITLE
Panos

### PR DESCRIPTION
**Pull request by Strae**
**Merge request checklist**

- [x] I read the [guidelines for contributing](https://github.com/mistic100/Photo-Sphere-Viewer/blob/master/.github/CONTRIBUTING.md)
- [x] I created my branch from `dev` and I am issuing the PR to `dev`
- [ ] I **didn't** commited files in the `dist` directory - I did, for Bower (your commend on my previous pull request...)
- [x] Unit tests are OK

This is a simple caching system for panorama; Browser's cache dont always work with xmlhttprequests, in this example I just clicked the "change Image" button lot of times:

without-cache
With cache:
with-cache

Now im working on some kinda of spooling process to avoid having a cache fairly big.

I also pushed the example directory, check out the index.htm for the example (you'll see 2 icons in the navbar, infinity for preload and nahil for clear cache)

### CODE REVIEW

**Comment by mistic100** [comment]
I didn't say to commit on "dist"... I said to not ignore and delete "dist" (what you did on your previous PR)

File: example/index.htm [hunk 1]
```diff
@@ -133,6 +151,7 @@
     time_anim: false,
     gyroscope: true,
     webgl: true,
+    usexmpdata: false,
```
Comment (mistic100) [inline example/index.htm]: whay changing the demo ?

File: src/js/PhotoSphereViewer.core.js [hunk 2]
```diff
@@ -74,21 +74,32 @@ PhotoSphereViewer.prototype._loadXMP = function() {
     throw new PSVError('Cannot load image');
   };
 
-  xhr.open('GET', this.config.panorama, true);
-  xhr.send(null);
-
+  if (true === this.prop.cacheTextures && 'undefined' != typeof this.prop.loadedXMP[this.config.panorama]) {
```
Comment (mistic100) [inline src/js/PhotoSphereViewer.core.js]: didn't you mixed "prop" and "config" ? seems like this is a user parameter => "config"

File: src/js/PhotoSphereViewer.public.js [hunk 3]
```diff
@@ -487,3 +487,45 @@ PhotoSphereViewer.prototype.startKeyboardControl = function() {
 PhotoSphereViewer.prototype.stopKeyboardControl = function() {
   window.removeEventListener('keydown', this);
 };
+
+/**
+ * Preload a panorama image and save it into internal cache.
+ * @param {String} pano - the file path
+ * @param {Function} callback - Progress callback, will receive the percentage as argument.
+ */
+PhotoSphereViewer.prototype.preloadPano = function(pano, callback) {
```
Comment (mistic100) [inline src/js/PhotoSphereViewer.public.js]: this callback a very little interest and is incoherent with the whole promise based API.
please remove it
+ missing the @returns annotation

File: src/js/PhotoSphereViewer.public.js [hunk 4]
```diff
@@ -487,3 +487,45 @@ PhotoSphereViewer.prototype.startKeyboardControl = function() {
...
+    return true;
+  }
+  return this._clearTexture(pano);
+};
+
+/**
+ * Return true if the panorama has been loaded.
+ * @param {string} the panorama file path.
+ */
+PhotoSphereViewer.prototype.panoHasLoaded = function(pano) {
```
Comment (mistic100) [inline src/js/PhotoSphereViewer.public.js]: => isPanoLoaded

File: src/js/PhotoSphereViewer.public.js [hunk 5]
```diff
@@ -487,3 +487,45 @@ PhotoSphereViewer.prototype.startKeyboardControl = function() {
...
+  return Boolean(this.prop.loadedTextures[pano]);
+};
+
+/**
+ * Return true if the panorama is in loading state.
+ * @param {string} the panorama file path.
+ */
+PhotoSphereViewer.prototype.panoIsLoading = function(pano) {
```
Comment (mistic100) [inline src/js/PhotoSphereViewer.public.js]: same remark as the progress callback

File: src/js/PhotoSphereViewer.js [hunk 6]
```diff
@@ -157,7 +157,12 @@ function PhotoSphereViewer(options) {
       cropped_height: 0,
       cropped_x: 0,
       cropped_y: 0
-    }
+    },
+    usexmpdata: false,
```
Comment (mistic100) [inline src/js/PhotoSphereViewer.js]: same remark as your previous PR : what is this ? aren't you mixing "prop" and "config"

**Comment by mistic100** [comment]
I'll wait for your todo to be resolved

File: example/index.htm (hunk 1 above)
Comment (Strae) [inline example/index.htm]: to show the preload functionality.. have I did something wrong?

Comment (mistic100) [inline example/index.htm]: I don't see the relation between disabling XMP parsing and the preload, and there shouln't be any.

File: src/js/PhotoSphereViewer.core.js (hunk 2 above)
Comment (Strae) [inline src/js/PhotoSphereViewer.core.js]: uop! you're right! i'll fix it

File: src/js/PhotoSphereViewer.public.js (hunk 3 above)
Comment (Strae) [inline src/js/PhotoSphereViewer.public.js]: Can't agree on that.
In my situation, I put markers on the scene to change panoramas (like a streetview tour) using the progress value to set the marker opacity.
If you dont like this feature, i'll keep it in my fork

File: src/js/PhotoSphereViewer.public.js [hunk 7]
```diff
@@ -487,3 +487,45 @@ PhotoSphereViewer.prototype.startKeyboardControl = function() {
...
+    return false;
+  }
+  return this._loadTexture(pano, progressCallback);
+};
+
+/**
+ * Remove a panorama image from the cache.
+ * @param {string} the file path
+ */
+PhotoSphereViewer.prototype.clearPano = function(pano) {
```
Comment (mistic100) [inline src/js/PhotoSphereViewer.public.js]: => clearCachedPanoramas

File: src/js/PhotoSphereViewer.public.js (hunk 4 above)
Comment (mistic100) [inline src/js/PhotoSphereViewer.public.js]: correction: => isPanoramaLoaded

File: src/js/PhotoSphereViewer.js (hunk 6 above)
Comment (Strae) [inline src/js/PhotoSphereViewer.js]: yeah, fixed.

File: src/js/PhotoSphereViewer.public.js (hunk 3 above)
Comment (mistic100) [inline src/js/PhotoSphereViewer.public.js]: Not that I don't like, but I want to keep to API clear and consistent. As promise/A+ spec does not include a progress event and this particular usage is not critical, I prefer to remove it.

File: src/js/PhotoSphereViewer.core.js [hunk 8]
```diff
@@ -74,8 +74,8 @@ PhotoSphereViewer.prototype._loadXMP = function() {
     throw new PSVError('Cannot load image');
   };
 
-  if (true === this.prop.cacheTextures && 'undefined' != typeof this.prop.loadedXMP[this.config.panorama]) {
-    defer.resolve(this.prop.loadedXMP[this.config.panorama]);
+  if (true === this.config.cacheTextures && 'undefined' != typeof this.config.loadedXMP[this.config.panorama]) {
+    defer.resolve(this.config.loadedXMP[this.config.panorama]);
```
Comment (mistic100) [inline src/js/PhotoSphereViewer.core.js]: Don't want to bother you but THIS has it's place in "prop"... same for "loadedTextures".
It's all about separate concerns (configurable vs internal)

Comment (Strae) [inline src/js/PhotoSphereViewer.core.js]: Yeah im a bit in a rush and did a quick copy'n'paste.
Now I need to work on another project, i'll do a proper update in the next days

File: src/js/PhotoSphereViewer.public.js (hunk 3 above)
Comment (Strae) [inline src/js/PhotoSphereViewer.public.js]: Ok, I understand and agree, but I think that this functionality can be useful for many.. i'll see how to implement this better.

Thanks for patience ;)

**Comment by Strae** [comment]
Created a simple cache registry and restored the original example/index.htm

The textures are complex object so it is very hard to state the memory usage of them, so I used the basic 'max number of items' approach

**Comment by mistic100** [comment]
You are really really complicating things. "cacheRegistry", "loadedTextures", "loadedXMP" and "_loadingTextures" are all the same thing. You can merge everything in a single array of objects :

```js
 {
  path,
  image,
  xmp,
  isLoading
}
```

"cacheTextures" and "maxCacheSize" can also be merged in a single param

File: src/js/PhotoSphereViewer.core.js [hunk 9]
```diff
@@ -446,16 +449,34 @@ PhotoSphereViewer.prototype._reverseAutorotate = function() {
...
   }
   return false;
 };
+
+/**
+ * Check if the cache size has reached the limit
+ * @private
+ * @return null
+ */
+PhotoSphereViewer.prototype._normalizeCacheSize = function(pano, progressCallback) {
```
Comment (mistic100) [inline src/js/PhotoSphereViewer.core.js]: This implementation is flawed, you remove only one item, whatever the number of extra items

Comment (mistic100) [inline src/js/PhotoSphereViewer.core.js]: "progressCallback" useless

File: src/js/PhotoSphereViewer.core.js [hunk 10]
```diff
@@ -156,8 +158,9 @@ PhotoSphereViewer.prototype._loadTexture = function(pano, progressCallback) {
       texture.minFilter = THREE.LinearFilter;
       texture.generateMipmaps = false;
       if (true === self.config.cacheTextures) {
-        self.config.loadedTextures[targetPano] = texture;
-        self.config._loadingTextures.splice(self.config._loadingTextures.indexOf(targetPano), 1);
+        self.prop.cacheRegistry.push(targetPano);
```
Comment (mistic100) [inline src/js/PhotoSphereViewer.core.js]: this should be in a separated method as well as lines 98-100

File: .gitignore [hunk 11]
```diff
@@ -3,3 +3,4 @@
 /.sass-cache
 /.idea
 /*.iml
+/dist
```
Comment (mistic100) [inline .gitignore]: Stop ignoring this

**Comment by mistic100** [comment]
Excuse me but this becomes a big mess. I will implement it myself.

**Comment by Strae** [comment]
Ok, no problem.
Please let me know when the functionality is ready, i'm really curios how you'll implement it!

thanks for patience ;)

**Comment by mistic100** [comment]
https://github.com/mistic100/Photo-Sphere-Viewer/commit/45e76eb6505b28fd1e677dc3106726306bb41792
It's the same thing but cleaner, not useless config and props, no promises mess *, handled in init and destroy.

----

* When I said that the API is promise based I didn't meant to put promises everywhere, putting promises on synchronous code like you did for "_normalizeCacheSize" and "_savePanoCache" is useless. Promises are really another way to deal with callbacks, you wouldn't add a callback to these methods.

**Comment by Strae** [comment]
Ok, seen your code, its really much cleaner.. 
The only thing that I need (but I guess can be handy for others too) is a way to tell the loading percentage of a panorama.
Instead of using callbacks, I was thinking to simply add a event, something like this (line 171 PhotoSphereViewer.core.js)
`
...
    var onprogress = function(e) {
      if (e.lengthComputable && self.loader) {
        var new_progress = parseInt(e.loaded / e.total * 100);
        if (new_progress > progress) {
          progress = new_progress;
          self.loader.setProgress(progress);
          self.trigger('panorama-progress', panorama, progress);
        }
      }
    };
...
`

**Comment by mistic100** [comment]
yes it's possible

**Comment by Strae** [comment]
ok, i'll add the event and make a pull request later
also fixed a couple of typos

jut to be clear, should I add the /dist folder into the pull or not?

**Comment by mistic100** [comment]
No, just don't touch it, and don't delete/ignore it either. Like I did in my commit.

**Pull request closed without merging**